retrofit/src/main/java/retrofit2/RequestFactory.java

```diff
@@ -680,6 +683,7 @@ private Headers parseHeaders(String[] headers) {
         return new ParameterHandler.PartMap<>(valueConverter, partMap.encoding());
 
       } else if (annotation instanceof Body) {
+        validateResolvableType(p, type);
```

Comment:
Having this in every branch sucks and is error-prone but we need to do the validation before running other code (like calling out to resolve converters)